File: packages/shared-schemas/src/today.ts
import { z } from "zod";

import { ModeSchema } from "./modes";

/**
 * Contrato del dashboard **Hoy** (build-plan Fase E): prioridades del día,
 * sugerencias proactivas y recap del día.
 *
 * Estado actual de los endpoints (junio 2026):
 *  - `/v1/tasks` → **operativo** (backend real, Tanda 1).
 *  - `/v1/suggestions` → **pendiente** (requiere LLM real + agenda; roadmap D2/F).
 *  - `/v1/recap` → **pendiente** (requiere LLM real; roadmap F). Hasta que
 *    existan, los hooks degradan a resultado vacío ante 404 (no error visible).
 *
 * Snake_case y `datetime({ offset: true })` para espejar la convención Pydantic
 * del resto del backend (FastAPI), de modo que al cablear el endpoint real el
 * shape ya coincida ("Pydantic gana, Zod sigue").
 */

// ---------- Prioridades del día (`GET /v1/tasks`) ----------

/** Estado de una tarea: pendiente o hecha (el check del wireframe 06). */
export const TaskStatusSchema = z.enum(["pending", "done"]);
export type TaskStatus = z.infer<typeof TaskStatusSchema>;

/**
 * Una prioridad del día. `scheduled_at` + `duration_min` arman la meta
 * "14:00 · 45 min"; cuando `status` es `done` la meta muestra "completada"
 * (derivado, no un campo aparte). El futuro modelo `Task` del backend sumará
 * campos de auditoría (created_at/updated_at) que no necesitamos en cliente.
 */
export const TaskSchema = z.object({
  id: z.string().uuid(),
  title: z.string().min(1),
  status: TaskStatusSchema,
  /** Hora agendada (ISO con offset). `null` si la tarea no tiene horario. */
  scheduled_at: z.string().datetime({ offset: true }).nullable(),
  /** Duración estimada en minutos. `null` si no aplica. */
  duration_min: z.number().int().positive().nullable(),
});
export type Task = z.infer<typeof TaskSchema>;

/** Respuesta de `GET /v1/tasks`: las prioridades del día + el total. */
export const TasksResponseSchema = z.object({
  items: z.array(TaskSchema),
  total: z.number().int().nonnegative(),
});
export type TasksResponse = z.infer<typeof TasksResponseSchema>;

/**
 * Body de `PATCH /v1/tasks/{id}`: por ahora sólo togglea el estado (marcar
 * hecha / re-abrir desde el check). Mirror de la rama mínima del futuro
 * `TaskUpdate`.
 */
export const TaskPatchSchema = z.object({
  status: TaskStatusSchema,
});
export type TaskPatch = z.infer<typeof TaskPatchSchema>;

// ---------- Sugerencias (`GET /v1/suggestions`) ----------

/**
 * Una sugerencia proactiva ("Ynara sugiere"): un título y su **porqué** (lo que
 * la hace honesta, no una orden arbitraria). `mode` la tinta y la asocia a un
 * modo; `null` si es transversal. Las genera el LLM real a futuro — hoy son
 * fixtures.
 */
export const SuggestionSchema = z.object({
  id: z.string().uuid(),
  title: z.string().min(1),
  /** El "porqué" de la sugerencia (subtítulo del wireframe 06/07). */
  why: z.string().min(1),
  /** Modo al que pertenece (para el tint), o `null` si es transversal. */
  mode: ModeSchema.nullable(),
});
export type Suggestion = z.infer<typeof SuggestionSchema>;

/** Respuesta de `GET /v1/suggestions`. */
export const SuggestionsResponseSchema = z.object({
  items: z.array(SuggestionSchema),
});
export type SuggestionsResponse = z.infer<typeof SuggestionsResponseSchema>;

// ---------- Recap del día (`GET /v1/recap`) ----------

/**
 * Recap del día (wireframe 15, CTA "Recap pendiente" del 06). `pending` mientras
 * no se cerró el día; al cerrarlo Ynara devuelve un `headline` editorial y los
 * `highlights` (lo que pasó). Generado por LLM a futuro; mock por ahora.
 */
export const RecapSchema = z.object({
  /** `true` mientras el día no se cerró (muestra el CTA en Hoy). */
  pending: z.boolean(),
  /** Fecha del día al que pertenece el recap (ISO con offset). */
  date: z.string().datetime({ offset: true }),
  /** Frase editorial de cierre (presente sólo si ya se generó). */
  headline: z.string().nullable(),
  /** Lo más importante del día, en bullets. Vacío si todavía está `pending`. */
  highlights: z.array(z.string()),
});
export type Recap = z.infer<typeof RecapSchema>;

File: apps/backend/app/schemas/today.py
"""Schemas del dashboard **Hoy** que aún no tenían backend: sugerencias + recap.

Espejo Pydantic del contrato del wire (``packages/shared-schemas/src/today.ts``,
``SuggestionSchema`` / ``SuggestionsResponseSchema`` / ``RecapSchema``): "Pydantic
gana, Zod sigue". Las prioridades del día (``/v1/tasks``) ya viven en
``app/schemas/task.py`` + ``task_api.py``; acá se completan las DOS superficies
que la web consumía contra mocks (``/v1/suggestions`` y ``/v1/recap``).

NO sagrado, NO es una tabla: estas respuestas se DERIVAN de las tareas del usuario
(``app/services/today.py``) —y ``suggestions`` además de sus prefs + memoria sembrada
para el cold-start (G5)—, no se persisten. La generación por LLM real (que la web
documenta como roadmap F) es la próxima fase; esta es la v1 derivada.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import Field

from app.enums import Mode
from app.schemas.base import YnaraBaseModel


class SuggestionOut(YnaraBaseModel):
    """Una sugerencia proactiva ("Ynara sugiere"). Espeja ``SuggestionSchema``.

    ``why`` es el "porqué" que la vuelve honesta (no una orden arbitraria); ``mode``
    la tiñe y la asocia a un modo, o ``None`` si es transversal. ``id`` es estable
    por sugerencia (la web lo usa como key de React): el servicio lo deriva
    determinísticamente del origen, no es aleatorio por request.
    """

    id: UUID
    title: str = Field(min_length=1)
    why: str = Field(min_length=1)
    mode: Mode | None


class SuggestionsResponse(YnaraBaseModel):
    """Respuesta de ``GET /v1/suggestions``. Espeja ``SuggestionsResponseSchema``.

    ``items`` vacío es válido y esperado (la web oculta la sección si no hay
    sugerencias): no se inventa contenido cuando no hay señal real.
    """

    items: list[SuggestionOut]


class RecapOut(YnaraBaseModel):
    """Recap del día. Espeja ``RecapSchema`` (wireframe 15, CTA del 06).

    - ``pending``: ``True`` cuando hay un borrador con contenido para mostrar (la web
      muestra el CTA "Recap pendiente" solo si ``pending``). En esta v1 sin
      mecanismo de cierre del día, ``pending`` modela "hay algo que recapitular":
      ``True`` si ``highlights`` no está vacío, ``False`` si no hubo actividad (así
      un usuario sin tareas no ve un CTA hacia un recap vacío).
    - ``date``: el día al que pertenece el recap (ISO con offset).
    - ``headline``: frase editorial de cierre. La generación real es por LLM (roadmap
      F); la v1 derivada arma una frase simple según la actividad, o ``None`` si no
      hay nada.
    - ``highlights``: lo más importante del día, en bullets, derivado de las tareas
      reales del usuario.
    """

    pending: bool
    date: datetime
    headline: str | None
    highlights: list[str]
